src/contact.c

```diff
@@ -98,6 +101,7 @@ int contact_add(struct contacts *contacts,
 	if (err)
 		goto out;
 
+	contact_remove(contacts, contact_find(contacts, c->uri));
```

Comment:
perhaps the contact handling could be moved to dictionary
and allowed for multiple entries ?

Reply:
I can't follow. This line comes from commit 010c2294fb27b78d3444137f6db6592eeb747760.

Reply:
if the number of contacts is very large, i.e. 1000 entries, this code might be slow.
If you are adding a new entry, it means that you have to search the whole list first.

In another follow-up PR we can consider to switch from `struct list` to `struct odict` which
is an ordered dictionary.

Reply:
Ok, I undestand. It's a hash table but for low bucket size it's like a list.

I can't see that odict is ordered or has any performance benefits compared to a hash map. But let's discuss this in a separate PR!